### PR TITLE
sysrepo-plugind FEATURE notification aging

### PR DESCRIPTION
**Pull request by Humblesaw**

### CODE REVIEW

File: src/executables/srpd_aging.c [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+	dir_st_t *opts = (dir_st_t *)arg;
+	char *arg1 = NULL, *arg2 = NULL, *remove_str = NULL, *notif_dir_name = NULL;
+	
+	notif_dir_name = get_notif_path();
+	if (!notif_dir_name) {
+		fprintf(stderr, "Fatal error: Notif directory is NULL\n");
+		goto cleanup;
+	}
+
+	while (opts->running) {
```
Comment (jktjkt) [inline src/executables/srpd_aging.c:143]: This looks like a busy wait loop. Why not sleep until it's time to perform an activity, and only then wake up?

File: CMakeLists.txt [hunk 2]
```diff
@@ -369,6 +371,18 @@ endif()
...
+find_program(ZIP_BINARY "zip")
+if(NOT ZIP_BINARY)
+    message(FATAL_ERROR "zip binary was not found.")
+endif()
+
+# mv
+find_program(MV_BINARY "mv")
+if(NOT MV_BINARY)
+    message(FATAL_ERROR "mv binary was not found.")
+endif()
```
Comment (jktjkt) [inline CMakeLists.txt:384]: Launching an external process to rename a file looks very, very fishy.

File: src/executables/sysrepo-plugind.c [hunk 3]
```diff
@@ -92,23 +95,6 @@ help_print(void)
...
-        sprintf(msg, "sysrepo-plugind error: %s\n", format);
-    } else {
-        sprintf(msg, "sysrepo-plugind error: %s (%s)\n", format, sr_strerror(sr_error));
-    }
-
-    va_start(ap, format);
-    vfprintf(stderr, msg, ap);
-    va_end(ap);
-}
-
```
Comment (jktjkt) [inline src/executables/sysrepo-plugind.c:111]: refactorings like this one belong to a separate commit

File: src/executables/sysrepo-plugind.c [hunk 4]
```diff
@@ -502,10 +488,7 @@ write_pidfile(int pidfd)
         return -1;
     }
 
-    if (snprintf(pid, sizeof(pid) - 1, "%ld\n", (long) getpid())) {
-        error_print(0, "Failed to allocate memory for pid (%s).", strerror(errno));
-        return -1;
-    }
```
Comment (jktjkt) [inline src/executables/sysrepo-plugind.c:508]: A patch about "notification aging" should not remove an error handler for a`snprtinf` in an unrelated function.

File: src/executables/sysrepo-plugind.c [hunk 5]
```diff
@@ -631,6 +616,20 @@ main(int argc, char **argv)
...
+        goto cleanup;
+    }
+
+    /* create notification rotation state data change subscription */
+    if ((r = sr_oper_get_subscribe(sess, "sysrepo-plugind", "/sysrepo-plugind:sysrepo-plugind/notif-datastore/rotation/archived-files-count",
+            sr_my_state_cb, &opts, 0, &subscr))) {
+        error_print(r, "Failed to subscribe");
+        goto cleanup;
+    }
+
```
Comment (jktjkt) [inline src/executables/sysrepo-plugind.c:632]: `sysrepo-plugin`'s purpose is to load *plugins* from a well-known path, and start executing these. Now, using `sysrepo-plugind` as an executor for this "notification aging" makes sense, but it should be a proper plugin, not something hardcoded in the executable's `main()`.

File: tests/CMakeLists.txt [hunk 6]
```diff
@@ -50,7 +50,8 @@ if(ENABLE_TESTS)
 
     # lists of all the tests
     set(tests test_modules test_context_change test_validation test_edit test_candidate test_operational test_lock
-        test_apply_changes test_copy_config test_rpc_action test_notif test_get test_process test_multi_connection test_nacm)
+        test_apply_changes test_copy_config test_rpc_action test_notif test_get test_process test_multi_connection test_aging 
+        test_nacm)
```
Comment (jktjkt) [inline tests/CMakeLists.txt:54]: nit: any reason for inserting a test in between the existing ones?

File: tests/test_aging.c [hunk 7]
```diff
@@ -0,0 +1,567 @@
...
+        if (errno != EEXIST) {
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+int
+remove_dir(char *path, int *test)
```
Comment (jktjkt) [inline tests/test_aging.c:79]: My advice is to rely on CMake's fixtures for setup and teardown. That way, this thing becomes a simple `rm -r` instead of 50 lines of C code.

The API is confusing -- it doesn't mention the fact that it's trying to work recursively, and the meaning of the `test` argument is not explained at all.

File: tests/test_aging.c [hunk 8]
```diff
@@ -0,0 +1,567 @@
...
+{
+    int rc = 0;
+    DIR *d;
+    struct dirent *dir;
+    char *new_path;
+
+    if (rmdir(path)) {
```
Comment (jktjkt) [inline tests/test_aging.c:86]: This should probably test for `ENOTEMPTY` because there's no error handling otherwise.

File: tests/test_aging.c [hunk 9]
```diff
@@ -0,0 +1,567 @@
...
+        d = opendir(path);
+	    if (!d) {
+	        fprintf(stderr, "Failed to open the directory in REMOVE_DIR\n");
+            return EXIT_FAILURE;
+	    }
+        while ((dir = readdir(d))) {
+
+			/** Skip current and parent directories **/
```
Comment (jktjkt) [inline tests/test_aging.c:94]: This does not skip current and parent directories, this skips all entries whose name begins with a dot.

File: tests/test_common.c [hunk 10]
```diff
@@ -36,8 +36,9 @@ _test_log_msg(sr_log_level_t level, const char *message, const char *prefix)
         severity = "WRN";
         break;
     case SR_LL_INF:
-        severity = "INF";
-        break;
+        /*severity = "INF";
+        break;*/
+        return;
```
Comment (jktjkt) [inline tests/test_common.c:41]: ?

File: tests/test_aging.c [hunk 11]
```diff
@@ -0,0 +1,567 @@
...
+
+    /** Store path to pidfile **/
+    if ((rc = asprintf(&data->pidfile, "%s%s", data->n_path, "my_pidfile")) == -1) {
+        goto cleanup;
+    }
+
+    /** Run archivation loop **/
+    data->pid = fork();
+    if (data->pid == 0) {
+        if (execl(SR_BINARY_DIR "/sysrepo-plugind", SR_BINARY_DIR "/sysrepo-plugind", "-d", "-p", data->pidfile, NULL)) {
```
Comment (jktjkt) [inline tests/test_aging.c:321]: This line starts `sysrepo-plugind` which will immediately start loading whatever plugins which are installed on the users' machine. That's not the expected behavior of a unit test.

File: src/executables/sysrepo-plugind.c (hunk 5 above)
Comment (michalvasko) [reply src/executables/sysrepo-plugind.c:632]: It seemed redundant to make it into a plugin but your are right that making it an "internal" plugin (that would not be dynamically loaded) would be a cleaner solution.

File: src/executables/srpd_common.c [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+#include <stdio.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+#include "bin_common.h"
+#include "srpd_aging.h"
+#include "sysrepo.h"
+
+struct srpd_int_plugin_s int_plugins[1] = {{aging_init_cb, aging_cleanup_cb, "aging_plugin"}};
```
Comment (michalvasko) [inline src/executables/srpd_common.c:32]: Doxygen comment.

File: src/executables/srpd_common.c [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+
+void
+srpd_error_print(int sr_error, const char *format, ...)
+{
+    va_list ap;
+    char msg[2048];
```
Comment (michalvasko) [inline src/executables/srpd_common.c:38]: Avoid fixed buffer length for messages.

File: src/executables/srpd_common.h [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+ * This source code is licensed under BSD 3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://opensource.org/licenses/BSD-3-Clause
+ */
+
+#include "sysrepo_types.h"
+
+#include <sys/types.h>
```
Comment (michalvasko) [inline src/executables/srpd_common.h:19]: Swap these includes, first system ones, then local headers.

File: src/executables/srpd_common.h [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+/**
+ * @brief Executes command given by @param cmd with arguments.
+ * First argument of the command has to be the command.
+ * See man ::execv(2).
+ *
+ * @param[in] cmd Command to be executed.
+ * @param[in] num_of_args Number of arguments given to function.
+ * @param[in] ... Arguments of the command.
+ * @return 0 on success.
+ * @return 1 on failure.
```
Comment (michalvasko) [inline src/executables/srpd_common.h:72]: Why this function returns `1` on error and the other ones `-1`?

File: tests/test_aging.c [hunk 16]
```diff
@@ -0,0 +1,532 @@
...
+    char *n_path;
+    char *a_path;
+    char *path_to_file;
+    char *pidfile;
+    uint32_t len_npath;
+    uint32_t len_apath;
+    pid_t pid;
+} test_data_t;
+
+/* Allocs memory, memory needs to be freed */
```
Comment (michalvasko) [inline tests/test_aging.c:55]: Redundant, understood from the return value type.

File: tests/test_aging.c [hunk 17]
```diff
@@ -0,0 +1,532 @@
...
+static char *
+get_test_path(char *which)
```
Comment (michalvasko) [inline tests/test_aging.c:57]: If `which` is not being modified, it should be `const`.

File: tests/test_aging.c [hunk 18]
```diff
@@ -0,0 +1,532 @@
...
+{
+	char *path;
+	if (asprintf(&path, "%s/data/%s", sr_get_repo_path(), which) == -1) {
```
Comment (michalvasko) [inline tests/test_aging.c:60]: Please leave a blank line after declarations.

File: tests/test_aging.c [hunk 19]
```diff
@@ -0,0 +1,532 @@
...
+		path = NULL;
+	}
+	return path;
```
Comment (michalvasko) [inline tests/test_aging.c:63]: Please leave a blank line before `return` unless some `out` parameters are not being also set there, for example.

File: tests/test_aging.c [hunk 20]
```diff
@@ -0,0 +1,532 @@
...
+}
+
+static int
+create_config(sr_session_ctx_t *sess, char *older_than, char *archive_dir, char *compress)
+{
+    int rc;
+    const struct ly_ctx *ctx;
+    struct lyd_node *node = NULL;
+
+    ctx = sr_session_acquire_context(sess);
```
Comment (michalvasko) [inline tests/test_aging.c:134]: On all errors you need to call `sr_session_release_context(sess);`.

File: tests/test_aging.c [hunk 21]
```diff
@@ -0,0 +1,532 @@
...
+            return EXIT_FAILURE;
+        }
+        if (sr_data) {
+            num = strtoul(lyd_get_value(lyd_child(lyd_child(lyd_child(sr_data->tree)))), NULL, 10);
+            sr_release_data(sr_data);
+        }
+        if (num >= archived_expected) {
+            break;
+        }
+        usleep(TIME_MAX);
```
Comment (michalvasko) [inline tests/test_aging.c:177]: `TIME_MAX` is not a suitable name for this macro.

**Review by michalvasko — CHANGES_REQUESTED** [review]

File: src/executables/srpd_aging.c (hunk 1 above)
Comment (Humblesaw) [reply src/executables/srpd_aging.c]: Fixed

File: modules/sysrepo-plugind.yang [hunk 22]
```diff
@@ -20,6 +24,38 @@ module sysrepo-plugind {
...
         }
 
+        container notif-datastore {
+            description "Includes configuration for notification datastore.";
+        
+            container rotation {
+                description "Notification rotation configuration and stats.";
+
+                container enabled {
+                    presence "Configuration is enabled.";
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply modules/sysrepo-plugind.yang:34]: Done

File: modules/sysrepo-plugind.yang [hunk 23]
```diff
@@ -20,6 +24,38 @@ module sysrepo-plugind {
...
+                    leaf older-than {
+                        description "Archive only notifications before configured time.";
+                        type string {
+                            pattern '[1-9][0-9]*[smhDWMY]';
+                        }
+                        mandatory true; 
+                    }
+                    leaf archive-dir {
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply modules/sysrepo-plugind.yang:42]: Done

File: modules/sysrepo-plugind.yang [hunk 24]
```diff
@@ -20,6 +24,38 @@ module sysrepo-plugind {
...
+                        type string;
+                        mandatory true;
+                    }
+                    leaf compress {
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply modules/sysrepo-plugind.yang:46]: Done

File: modules/sysrepo-plugind.yang [hunk 25]
```diff
@@ -20,6 +24,38 @@ module sysrepo-plugind {
...
+                        type boolean;
+                        default "true";
+                    }
+                }
+
+                leaf archived-files-count {
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply modules/sysrepo-plugind.yang:52]: Done

File: src/executables/srpd_aging.c [hunk 26]
```diff
@@ -0,0 +1,377 @@
...
+#include "bin_common.h"
+#include "config.h"
+#include "srpd_aging.h"
+#include "srpd_common.h"
+
+int
+aging_init_cb(sr_session_ctx_t *session, void **private_data)
+{
+    int r = 0;
+    dir_st_t *opts = (dir_st_t *)calloc(1, sizeof(dir_st_t));
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply src/executables/srpd_aging.c:39]: Done

File: src/executables/srpd_aging.c [hunk 27]
```diff
@@ -0,0 +1,377 @@
...
+
+    opts->subscr = NULL;
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply src/executables/srpd_aging.c:41]: Done

File: src/executables/srpd_aging.c [hunk 28]
```diff
@@ -0,0 +1,377 @@
...
+    if ((r = sr_oper_get_subscribe(session, "sysrepo-plugind", "/sysrepo-plugind:sysrepo-plugind/notif-datastore/rotation/archived-files-count",
+            sr_my_state_cb, opts, 0, &opts->subscr))) {
+        srpd_error_print(r, "Failed to subscribe");
+        goto cleanup;
+    }
+
+    *private_data = opts;
+    return 0;
+
+cleanup:
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply src/executables/srpd_aging.c:60]: Renamed to error

File: src/executables/srpd_aging.c [hunk 29]
```diff
@@ -0,0 +1,377 @@
...
+    dir_st_t *opts = private_data;
+
+    sr_unsubscribe(opts->subscr);
+    opts->finish = 1;
+    sr_my_notif_cb(session, 0, NULL, NULL, 0, 0, opts);
+    free(opts);
+}
+
+static char *
+get_notif_path(void)
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply src/executables/srpd_aging.c:78]: Done

File: src/executables/srpd_aging.c [hunk 30]
```diff
@@ -0,0 +1,377 @@
...
+    } else {
+        if (asprintf(&path, "%s/data/notif/", sr_get_repo_path()) == -1) {
+            path = NULL;
+        }
+    }
+    return path;
+}
+
+int
+sr_my_state_cb(sr_session_ctx_t *session, uint32_t sub_id, const char *module_name, const char *path,
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply src/executables/srpd_aging.c:93]: Renamed

File: src/executables/srpd_aging.c [hunk 31]
```diff
@@ -0,0 +1,377 @@
...
+        goto cleanup;
+    }
+
+cleanup:
+    sr_session_release_context(session);
+    return rc;
+}
+
+static int
+format_control(char *string, time_t *file_time1, time_t *file_time2)
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply src/executables/srpd_aging.c:120]: Added + avoided

File: src/executables/srpd_aging.c [hunk 32]
```diff
@@ -0,0 +1,377 @@
...
+                if (!rc && opts->compress) {
+                    if (asprintf(&remove_str, "%s%s", notif_dir_name, dir->d_name) == -1) {
+                        goto cleanup;
+                    }
+
+                    /* Remove a file from notif folder */
+                    if (remove(remove_str)) {
+                        srpd_error_print(0, "Removing a file failed.");
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply src/executables/srpd_aging.c:253]: Done

File: src/executables/srpd_aging.c [hunk 33]
```diff
@@ -0,0 +1,377 @@
...
+    free(arg2);
+    free(remove_str);
+    free(notif_dir_name);
+    free(opts->archive);
+    closedir(d);
+    return NULL;
+}
+
+int
+sr_my_notif_cb(sr_session_ctx_t *session, uint32_t sub_id, const char *module_name, const char *xpath,
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply src/executables/srpd_aging.c:280]: Renamed

File: src/executables/srpd_aging.c [hunk 34]
```diff
@@ -0,0 +1,377 @@
...
+    (void) event;
+    (void) request_id;
+
+    if (opts->finish) {
+        opts->running = 0;
+        pthread_join(opts->tid, NULL);
+        return SR_ERR_OK;
+    }
+
+    sr_get_changes_iter(session, "/sysrepo-plugind:sysrepo-plugind/notif-datastore/rotation/enabled//.", &iter);
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply src/executables/srpd_aging.c:303]: Check created

File: src/executables/srpd_aging.c [hunk 35]
```diff
@@ -0,0 +1,377 @@
...
+            /* fallthrough */
+            case 'h':                           /* time in hours */
+                time_value = time_value * 60;
+            /* fallthrough */
+            case 'm':                           /* time in minutes */
+                time_value = time_value * 60;
+            /* fallthrough */
+            case 's':                           /* time in seconds */
+                break;
+            }
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply src/executables/srpd_aging.c:345]: Correct, converting is done progressively (days to hours, then hours to minutes, etc)

File: src/executables/srpd_aging.c [hunk 36]
```diff
@@ -0,0 +1,377 @@
...
+        }
+    }
+    sr_free_change_iter(iter);
+
+    /* check whether a thread should be created */
+    if (t_creat) {
+        assert(opts->archive);
+        opts->running = 1;
+        if ((rc = pthread_create(&(opts->tid), NULL, &sr_archive_loop, opts))) {
+            goto t_error;
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply src/executables/srpd_aging.c:368]: Label removed

File: src/executables/srpd_aging.h [hunk 37]
```diff
@@ -0,0 +1,58 @@
...
+typedef struct {
+    time_t config_time;
+    char *archive;
+    int compress;
+    uint64_t archived_notif_count;
+    sr_subscription_ctx_t *subscr;
+    pthread_t tid;
+    int running;
+    int finish;
+} dir_st_t;
```
[the remark this answers is not in the document]
Comment (Humblesaw) [reply src/executables/srpd_aging.h:32]: Renamed

File: src/executables/srpd_common.c (hunk 12 above)
Comment (Humblesaw) [reply src/executables/srpd_common.c:32]: Done

File: src/executables/srpd_common.c (hunk 13 above)
Comment (Humblesaw) [reply src/executables/srpd_common.c:38]: Error print moved back to sysrepo-plugind.c

File: src/executables/srpd_common.h (hunk 14 above)
Comment (Humblesaw) [reply src/executables/srpd_common.h:19]: Done

File: src/executables/srpd_common.h (hunk 15 above)
Comment (Humblesaw) [reply src/executables/srpd_common.h:72]: Changed to -1

File: tests/test_aging.c (hunk 16 above)
Comment (Humblesaw) [reply tests/test_aging.c:55]: Removed

File: tests/test_aging.c (hunk 17 above)
Comment (Humblesaw) [reply tests/test_aging.c:57]: Done

File: tests/test_aging.c (hunk 19 above)
Comment (Humblesaw) [reply tests/test_aging.c:63]: Done

File: tests/test_aging.c (hunk 20 above)
Comment (Humblesaw) [reply tests/test_aging.c:134]: FIxed

File: tests/test_aging.c (hunk 21 above)
Comment (Humblesaw) [reply tests/test_aging.c:177]: Renamed

File: src/executables/srpd_aging.c (hunk 34 above)
[the remark this answers is not in the document]
Comment (michalvasko) [reply src/executables/srpd_aging.c:333]: Return value still not checked.

File: src/executables/srpd_aging.c [hunk 38]
```diff
@@ -0,0 +1,403 @@
...
+
+    sr_get_changes_iter(session, "/sysrepo-plugind:sysrepo-plugind/notif-datastore/rotation/enabled//.", &iter);
+
+    while ((rc = sr_get_change_tree_next(session, iter, &oper, &node, NULL, NULL, NULL)) == SR_ERR_OK) {
+
+        if (!strcmp(node->schema->name, "enabled")) {
+            if (oper == SR_OP_CREATED) {
+                t_creat = 1;
+            } else if (oper == SR_OP_DELETED) {
+                rc = srpd_join_thread(opts);
```
Comment (michalvasko) [inline src/executables/srpd_aging.c:341]: If continuing after an error, at least add a comment explaining that it is intentional.

**Review by michalvasko — CHANGES_REQUESTED** [review]

File: src/executables/srpd_aging.c (hunk 34 above)
[the remark this answers is not in the document]
Comment (Humblesaw) [reply src/executables/srpd_aging.c:333]: Done

File: src/executables/srpd_aging.c (hunk 38 above)
Comment (Humblesaw) [reply src/executables/srpd_aging.c:341]: Done

File: src/executables/srpd_rotation.c [hunk 39]
```diff
@@ -0,0 +1,409 @@
...
+    opts->running = 0;
+    if ((rc = pthread_join(opts->tid, NULL))) {
+        SRPLG_LOG_ERR("srpd_rotation", "pthread_join failed (%s).", sr_strerror(rc));
+    }
+
+    return rc;
+}
+
+int
+srpd_notif_cb(sr_session_ctx_t *session, uint32_t sub_id, const char *module_name, const char *xpath,
```
Comment (michalvasko) [inline src/executables/srpd_rotation.c:312]: Generic name, use something like `srpd_rotation_change_cb`.

File: src/executables/srpd_rotation.c [hunk 40]
```diff
@@ -0,0 +1,409 @@
...
+}
+
+void
+srpd_rotation_cleanup_cb(sr_session_ctx_t *session, void *private_data)
+{
+    srpd_rotation_opts_t *opts = (srpd_rotation_opts_t *)private_data;
+
+    sr_unsubscribe(opts->subscr);
+    opts->finish = 1;
+    srpd_notif_cb(session, 0, NULL, NULL, 0, 0, opts);
```
Comment (michalvasko) [inline src/executables/srpd_rotation.c:73]: Why delegate thread join to the change callback and not perform it in this function?

File: tests/test_rotation.c [hunk 41]
```diff
@@ -0,0 +1,535 @@
...
+#include <cmocka.h>
+
+#include <libyang/libyang.h>
+
+#include "sysrepo.h"
+#include "sysrepo_types.h"
+#include "test_common.h"
+#include "config.h"
+
+#define MAX_WAIT_MICROSEC 100000
```
Comment (michalvasko) [inline tests/test_rotation.c:40]: I did not like the name because `MAX_WAIT` says nothing about where this value is used. Something like `TIMEOUT_STEP_MS` would be better.

**Review by michalvasko — CHANGES_REQUESTED** [review]

File: tests/test_rotation.c (hunk 41 above)
Comment (Humblesaw) [reply tests/test_rotation.c:40]: Changed

File: src/executables/srpd_rotation.c (hunk 40 above)
Comment (Humblesaw) [reply src/executables/srpd_rotation.c:73]: True, I changed it

File: src/executables/srpd_rotation.c (hunk 39 above)
Comment (Humblesaw) [reply src/executables/srpd_rotation.c:312]: Done

File: src/executables/srpd_rotation.h [hunk 42]
```diff
@@ -0,0 +1,57 @@
...
+ */
+typedef struct {
+    time_t rotation_time;
+    char *output_folder;
+    int compress;
+    uint64_t rotated_files_count;
+    sr_subscription_ctx_t *subscr;
+    pthread_t tid;
+    int running;
+} srpd_rotation_opts_t;
```
Comment (michalvasko) [inline src/executables/srpd_rotation.h:31]: Any members accessed from several threads need to be atomic/lock-protected.

File: tests/test_rotation.c [hunk 43]
```diff
@@ -0,0 +1,535 @@
...
+#include <cmocka.h>
+
+#include <libyang/libyang.h>
+
+#include "sysrepo.h"
+#include "sysrepo_types.h"
+#include "test_common.h"
+#include "config.h"
+
+#define TIMEOUT_STEP_MS 100000
```
Comment (michalvasko) [inline tests/test_rotation.c:40]: Sorry, microseconds are `us`, milliseconds are `ms`.

**Review by michalvasko — CHANGES_REQUESTED** [review]

File: src/executables/srpd_rotation.c [hunk 44]
```diff
@@ -0,0 +1,398 @@
...
+#include "srpd_common.h"
+#include "srpd_rotation.h"
+
+int
+srpd_rotation_init_cb(sr_session_ctx_t *session, void **private_data)
+{
+    int r = 0;
+    srpd_rotation_opts_t *opts;
+
+    opts = calloc(1, sizeof *opts);
```
Comment (michalvasko) [inline src/executables/srpd_rotation.c:41]: Check return value.

File: src/executables/srpd_rotation.c [hunk 45]
```diff
@@ -0,0 +1,398 @@
...
+    return r;
+}
+
+void
+srpd_rotation_cleanup_cb(sr_session_ctx_t *session, void *private_data)
+{
+    int rc;
+    srpd_rotation_opts_t *opts = (srpd_rotation_opts_t *)private_data;
+
+    (void) session;
```
Comment (michalvasko) [inline src/executables/srpd_rotation.c:72]: Redundant space.

File: src/executables/srpd_rotation.c [hunk 46]
```diff
@@ -0,0 +1,398 @@
...
+    int rc = 0;
+    const struct ly_ctx *ctx;
+    char value[20];
+    srpd_rotation_opts_t *opts = (srpd_rotation_opts_t *)private_data;
+
+    (void) sub_id;
+    (void) module_name;
+    (void) path;
+    (void) request_xpath;
+    (void) request_id;
```
Comment (michalvasko) [inline src/executables/srpd_rotation.c:116]: Redundant spaces.

File: src/executables/srpd_rotation.c [hunk 47]
```diff
@@ -0,0 +1,398 @@
...
+    int rc = 0;
+    DIR *d = NULL;
+    struct dirent *dir = NULL;
+    time_t current_time, file_time2 = 0;
+    srpd_rotation_opts_t *opts = (srpd_rotation_opts_t *)arg;
+    char *arg1 = NULL, *arg2 = NULL, *remove_str = NULL, *notif_dir_name = NULL;
+
+    notif_dir_name = srpd_get_notif_path();
+    if (!notif_dir_name) {
+        SRPLG_LOG_ERR("srpd_rotation", "Fatal error: Notif directory is NULL.");
```
Comment (michalvasko) [inline src/executables/srpd_rotation.c:195]: Redundant `Fatal error`, it is obvious from the message severity.

File: src/executables/srpd_rotation.c [hunk 48]
```diff
@@ -0,0 +1,398 @@
...
+        goto cleanup;
+    }
+    if (srpd_mkpath((char *)ATOMIC_PTR_LOAD_RELAXED(opts->output_folder), 0777) == -1) {
+        SRPLG_LOG_ERR("srpd_rotation", "Fatal error: Archive directory could not be created");
```
Comment (michalvasko) [inline src/executables/srpd_rotation.c:199]: Redundant `Fatal error`, it is obvious from the message severity.

File: src/executables/srpd_rotation.c [hunk 49]
```diff
@@ -0,0 +1,398 @@
...
+                    } else {
+                        ATOMIC_INC_RELAXED(opts->rotated_files_count);
+
+                        if (asprintf(&remove_str, "%s%s", notif_dir_name, dir->d_name) == -1) {
+                            goto cleanup;
+                        }
+
+                        /* Remove a file from notif folder */
+                        if (remove(remove_str)) {
+                            SRPLG_LOG_ERR("srpd_rotation", "Removing a file failed.");
```
Comment (michalvasko) [inline src/executables/srpd_rotation.c:252]: Include file name in the error.

File: src/executables/srpd_rotation.c [hunk 50]
```diff
@@ -0,0 +1,398 @@
...
+                    if (asprintf(&arg1, "%s%s", notif_dir_name, dir->d_name) == -1) {
+                        goto cleanup;
+                    }
+                    if (asprintf(&arg2, "%s%s", (char *)ATOMIC_PTR_LOAD_RELAXED(opts->output_folder), dir->d_name) == -1) {
+                        goto cleanup;
+                    }
+
+                    /* Move a file to the output folder */
+                    if ((rc = rename(arg1, arg2)) == -1) {
+                        SRPLG_LOG_ERR("srpd_rotation", "Moving a file failed.");
```
Comment (michalvasko) [inline src/executables/srpd_rotation.c:267]: Include the moved file name in the error.

File: src/executables/srpd_rotation.c [hunk 51]
```diff
@@ -0,0 +1,398 @@
...
+        d = NULL;
+    }
+
+cleanup:
+    free(arg1);
+    free(arg2);
+    free(remove_str);
+    free(notif_dir_name);
+    free(ATOMIC_PTR_LOAD_RELAXED(opts->output_folder));
+    ATOMIC_PTR_STORE_RELAXED(opts->output_folder, NULL);
```
Comment (michalvasko) [inline src/executables/srpd_rotation.c:293]: Why is the folder being freed here?

File: src/executables/srpd_rotation.c [hunk 52]
```diff
@@ -0,0 +1,398 @@
...
+    sr_change_oper_t oper;
+    const struct lyd_node *node;
+    srpd_rotation_opts_t *opts = (srpd_rotation_opts_t *)private_data;
+    void *temp = NULL;
+
+    (void) sub_id;
+    (void) module_name;
+    (void) xpath;
+    (void) event;
+    (void) request_id;
```
Comment (michalvasko) [inline src/executables/srpd_rotation.c:315]: Redundant spaces.

File: src/executables/srpd_rotation.c [hunk 53]
```diff
@@ -0,0 +1,398 @@
...
+    srpd_rotation_opts_t *opts = (srpd_rotation_opts_t *)private_data;
+
+    (void) session;
+
+    sr_unsubscribe(opts->subscr);
+    ATOMIC_STORE_RELAXED(opts->running, 0);
+    if ((rc = pthread_join(opts->tid, NULL))) {
+        SRPLG_LOG_ERR("srpd_rotation", "pthread_join failed (%s).", rc);
+    }
+    free(opts);
```
Comment (michalvasko) [inline src/executables/srpd_rotation.c:79]: Free the output folder here as well.

**Review by michalvasko — CHANGES_REQUESTED** [review]

File: src/executables/srpd_rotation.c (hunk 44 above)
Comment (Humblesaw) [reply src/executables/srpd_rotation.c:41]: Done

File: src/executables/srpd_rotation.c (hunk 45 above)
Comment (Humblesaw) [reply src/executables/srpd_rotation.c:72]: Done

File: src/executables/srpd_rotation.c (hunk 46 above)
Comment (Humblesaw) [reply src/executables/srpd_rotation.c:116]: Done

File: src/executables/srpd_rotation.c (hunk 47 above)
Comment (Humblesaw) [reply src/executables/srpd_rotation.c:195]: Fixed

File: src/executables/srpd_rotation.c (hunk 48 above)
Comment (Humblesaw) [reply src/executables/srpd_rotation.c:199]: Fixed

File: src/executables/srpd_rotation.c (hunk 49 above)
Comment (Humblesaw) [reply src/executables/srpd_rotation.c:252]: Done

File: src/executables/srpd_rotation.c (hunk 50 above)
Comment (Humblesaw) [reply src/executables/srpd_rotation.c:267]: Done

File: src/executables/srpd_rotation.c (hunk 51 above)
Comment (Humblesaw) [reply src/executables/srpd_rotation.c:293]: Done

File: src/executables/srpd_rotation.c (hunk 52 above)
Comment (Humblesaw) [reply src/executables/srpd_rotation.c:315]: Done

File: src/executables/srpd_rotation.c (hunk 53 above)
Comment (Humblesaw) [reply src/executables/srpd_rotation.c:79]: Done

**Review by michalvasko — APPROVED** [review]

File: src/executables/sysrepo-plugind.c [hunk 54]
```diff
@@ -164,6 +156,33 @@ handle_signals(void)
...
+    /* get plugins dir from environment variable, or use default one */
+    *plugins_dir = getenv("SRPD_PLUGINS_PATH");
+    if (!*plugins_dir) {
+        *plugins_dir = SRPD_PLG_PATH;
+    }
+
+    /* create the directory if it does not exist */
+    if (access(*plugins_dir, F_OK) == -1) {
+        if ((errno != ENOENT) && (errno != EACCES)) {
```
Comment (michalvasko) [inline src/executables/sysrepo-plugind.c:172]: `EACCESS` should be an error.

**Review by michalvasko — CHANGES_REQUESTED** [review]

**Review by michalvasko — APPROVED** [review]
